### PR TITLE
Polars: test pickle bombs

### LINKED ISSUE #1840: Polars: test pickle bombs
We should have CI tests that ensure that manipulation of LazyFrames never triggers pickle deserialization. If this ever changes in Polars, we'll then be alerted by failing CI.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #1840

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#1841** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1841?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python/test/test_polars.py [hunk 1]
```diff
@@ -548,3 +548,69 @@ def test_replace_binary_path():
     assert str(m_expr((pl.LazyFrame(dict()), pl.all()))) == "len().testing!:noise_plugin()"
 
     del os.environ["OPENDP_LIB_PATH"]
+
+
+def test_pickle_bomb():
+    pl = pytest.importorskip("polars")
+
+    from polars._utils.parse import parse_into_list_of_expressions  # type: ignore[import-not-found]
```
Comment (mccalluc) [inline python/test/test_polars.py:556]:
```suggestion
    from polars._utils.parse import parse_into_list_of_expressions  # type: ignore[import-not-found]
    from polars._utils.wrap import wrap_expr
```
Do all the imports at the top?

I'm wary of the imports from `_utils`: Definitely wouldn't rely on this outside of tests, but maybe if it helps us write a clean test, it's worth the risk? Any way we could use a higher-level, public interface?

File: python/test/test_polars.py [hunk 2]
```diff
@@ -548,3 +548,69 @@ def test_replace_binary_path():
...
+
+    poison_binary = b'c__builtin__\neval\n(V1 / 0\ntR.'
+
+    # poison_binary raises a ZeroDivisionError if it is ever unpickled
+    with pytest.raises(ZeroDivisionError):
+        pickle.loads(poison_binary)
+
+    # craft an expression that contains a poisoned pickle binary
+    from polars._utils.wrap import wrap_expr
+
```
Comment (mccalluc) [inline python/test/test_polars.py:569]:
```suggestion
```

File: python/test/test_polars.py [hunk 3]
```diff
@@ -548,3 +548,69 @@ def test_replace_binary_path():
...
+    pyexprs = parse_into_list_of_expressions((pl.len(), pl.lit("Laplace"), pl.lit(0.7)))
+    bomb_expr = wrap_expr(
+        pl.polars.register_plugin_function(
+            plugin_path=str(lib_path),
+            function_name="noise",
+            args=pyexprs,
```
Comment (mccalluc) [inline python/test/test_polars.py:575]:
```suggestion
    py_exprs = parse_into_list_of_expressions((pl.len(), pl.lit("Laplace"), pl.lit(0.7)))
    bomb_expr = wrap_expr(
        pl.polars.register_plugin_function(
            plugin_path=str(lib_path),
            function_name="noise",
            args=py_exprs,
```

File: rust/src/polars/mod.rs [hunk 4]
```diff
@@ -57,6 +57,10 @@ where
                 return Ok(None);
             }
 
+            if !kwargs.is_empty() {
+                return fallible!(FailedFunction, "OpenDP does not allow pickled keyword arguments as they may enable remote code execution.");
```
Comment (mccalluc) [inline rust/src/polars/mod.rs:61]: My understanding is that pickle is specific to python, but polars also supports R: If any eventual R code might also exercise this path, should the error message be more language-agnostic?

File: python/test/test_polars.py [hunk 5]
```diff
@@ -548,3 +548,69 @@ def test_replace_binary_path():
...
+    assert err.value.message == err_msg
+
+    with pytest.raises(dp.OpenDPException) as err:
+        dp.m.make_private_lazyframe(
+            lf_domain,
+            dp.symmetric_distance(),
+            dp.max_divergence(T=float),
+            bomb_lf,
+        )
+    assert err.value.message == err_msg
```
Comment (mccalluc) [inline python/test/test_polars.py:616]: use the `match` kwarg?
```suggestion
    err_msg_re = re.escape("OpenDP does not allow pickled keyword arguments as they may enable remote code execution.")
    with pytest.raises(dp.OpenDPException, match=err_msg_re) as err:
        dp.m.make_private_expr(
            dp.expr_domain(lf_domain, grouping_columns=[]),
            dp.partition_distance(dp.symmetric_distance()),
            dp.max_divergence(T=float),
            bomb_expr,
        )

    with pytest.raises(dp.OpenDPException, match=err_msg_re) as err:
        dp.m.make_private_lazyframe(
            lf_domain,
            dp.symmetric_distance(),
            dp.max_divergence(T=float),
            bomb_lf,
        )
```

**Review by mccalluc — COMMENTED** [review]
A couple little tweaks. Only real concern is with using non-public methods, but they might be the best option here?

File: python/test/test_polars.py [hunk 6]
```diff
@@ -548,3 +548,69 @@ def test_replace_binary_path():
...
+
+def test_pickle_bomb():
+    pl = pytest.importorskip("polars")
+
+    from polars._utils.parse import parse_into_list_of_expressions  # type: ignore[import-not-found]
+    from opendp._lib import lib_path
+    import io
+    import pickle
+
+    poison_binary = b'c__builtin__\neval\n(V1 / 0\ntR.'
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply python/test/test_polars.py:561]: I'm not sure how I would produce this with pickle.dumps.

File: python/test/test_polars.py (hunk 1 above)
Comment (Shoeboxam) [reply python/test/test_polars.py:556]: To use a higher-level Polars API, I'd have to further reverse-engineer the pickle protocol. I'd rather not spend more time on that, when this is so simple already.

File: python/test/test_polars.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply python/test/test_polars.py:561]: Left a comment to a blog post!

File: rust/src/polars/mod.rs (hunk 4 above)
Comment (Shoeboxam) [reply rust/src/polars/mod.rs:61]: Polars doesn't directly support plugins from R at this point, but if you know how things work, you can still write those DSL nodes for plugins into expressions in R. Now that we don't use pickling, R will be able to fill kwargs fields in these new DSL nodes with empty strings.


Ultimately, the slot to fill in these DSL nodes is binary pickle data, so if someone were to craft a DSL node with data in there from R, the error is still appropriate.

**Review by mccalluc — APPROVED** [review]
Thanks! Sorry for the quibbles.

**Review by damienbfs — APPROVED** [review]
Thanks a lot for adding the test!

**Pull request merged**